R/Module-PatientLevelPredictionValidation.R

```diff
@@ -0,0 +1,254 @@
+# PatientLevelPredictionValidationModule -------------
+#' @title Module for performing validation of patient-level prediction models
+#' @export
+#' @description
+#' Module for performing patient-level prediction model validation for models
+#' built using the PatientLevelPrediction package.
+PatientLevelPredictionValidationModule <- R6::R6Class(
+  classname = "PatientLevelPredictionValidationModule",
+  inherit = StrategusModule,
+  public = list(
+    #' @field tablePrefix The table prefix to append to the results tables
+    tablePrefix = "val_",
+    #' @description Initialize the module
+    initialize = function() {
+      super$initialize()
+    },
+    #' @description Executes the PatientLevelPrediction package to validate a
+    #' PLP model
+    #' @template connectionDetails
+    #' @template analysisSpecifications
+    #' @template executionSettings
+    execute = function(connectionDetails, analysisSpecifications, executionSettings) {
+      super$.validateCdmExecutionSettings(executionSettings)
+      super$execute(connectionDetails, analysisSpecifications, executionSettings)
+
+      jobContext <- private$jobContext
+      #cohortDefinitionSet <- super$.createCohortDefinitionSetFromJobContext()
+      workFolder <- jobContext$moduleExecutionSettings$workSubFolder
+      resultsFolder <- jobContext$moduleExecutionSettings$resultsSubFolder
+
+      #library(PatientLevelPrediction)
+      private$.message("Validating inputs")
+      inherits(jobContext, 'list')
+
+      if (is.null(jobContext$settings)) {
+        stop("Analysis settings not found in job context")
+      }
+      # if (is.null(jobContext$sharedResources)) {
+      #   stop("Shared resources not found in job context")
+      # }
+      if (is.null(jobContext$moduleExecutionSettings)) {
+        stop("Execution settings not found in job context")
+      }
+
+      #workFolder <- jobContext$moduleExecutionSettings$workSubFolder
+      #resultsFolder <- jobContext$moduleExecutionSettings$resultsSubFolder
+
+      private$.message("Executing PLP Validation")
+      #moduleInfo <- getModuleInfo()
+
+      # find where cohortDefinitions are as sharedResources is a list
+      # cohortDefinitionSet <- createCohortDefinitionSetFromJobContext(
+      #   sharedResources = jobContext$sharedResources,
+      #   settings = jobContext$settings
+      # )
+
+      # check the model locations are valid and apply model
+      upperWorkDir <- dirname(jobContext$moduleExecutionSettings$workFolder) # AGS: NOTE - Using the "root" folder as the expection is that the ModelTransferModule output is here
```

Comment:
Linking this to #21 since I'm unsure of the workflow between the ModelTransferModule and the PLP Validation module. In looking at the `upperWorkDir` the assumption I see is that the ModelTransferModule is written to the "root" folder that holds the results/work sub folders. Is this correct?

Reply:
I'm not sure I completely follow. The `ModelTransferModule` should write the models to either results/work module subfolder in a directory `models`. The current code writes it in the work subfolder. Then it writes a csv file in the results directory that has the paths to each individual model in the work directory. 

So something like:

```text
├── strategusOutput
│   ├── DatabaseMetaData
|   ├── CohortGeneratorModule
│   ├── DatabaseMetaData
│   ├── ModelTransferModule
│   │   └── github_export.csv
├── strategusWork
│   ├── CohortGeneratorModule
│   ├── ModelTransferModule
│   │   └── models
│   │       └── model_github_PandemicPrediction_master
│   │           ├── model_1
│   │           ├── model_2
```

Since the downloaded models are the results of the `ModelTransferModule` it could be argued they should be in the results subfolder instead (`strategusOutput/ModelTransferModule/models`) . But I would defer to you which fits better in with Strategus. 

In current code the `upperWorkDir` should be the `strategusWork` directory. And in line 61 `modelSavelocation` should point to `strategusWork/ModelTransferModule/models` and that's where the validation module looks for the models. I'm using `sort(dir(upperWorkDir, pattern = 'ModelTransferModule'), decreasing = T)[1]` to identify the correct module since it used to have trailing underscores and numbers. If that's not the case anymore this can be simplified to a simple `file.path`.

Hope that's clear! 
Egill